### PR TITLE
recipes - add raster-vector function, clean logic in graph funcs

### PR DESCRIPTION
**Pull request by KrisCowie**
Added a raster-vector function that will take in the first layer of a raster geotiff, and aggregate the values to a given list of polygons. I.e. given a population geotiff from here for example https://ghsl.jrc.ec.europa.eu/download.php?ds=pop , and a geodataframe which contains the geometries of FSAs, it will aggregate the population and add a column to the gdf with the summed pop for each polygon.
Cleaned up some logic in the graphing functions that was breaking in edge cases.
Finally, added importlib to the scaling function so users can enter whichever sklearn.preprocessing module they like, and it will still work, without having to account for all of them manually (unfortunately, the same trick doesn't work with the models, as they come from too many different packages).

### CODE REVIEW

File: recipes/categorical_anomaly.py [hunk 1]
```diff
@@ -269,15 +342,23 @@ def graph_scatter(self, **kwargs):
 
 
     def graph_geo(self, **kwargs):
+        """
+        If a uses provides a geodataframe with applicable geometries, this function maps those using Folium.
+
+        """
         if type(self.metrics_column) is str:
-            hover = [self.metrics_column]
+            hover = [self.metrics_column].copy()
```
Comment (woozyking) [inline recipes/categorical_anomaly.py]: copy shouldn't be necessary as a new list is created anyway

File: recipes/categorical_anomaly.py [hunk 2]
```diff
@@ -5,18 +5,23 @@
 import geopandas as gpd
 import tqdm
 
+import rasterio
+import rasterio.mask
+from rasterio import open
```
Comment (woozyking) [inline recipes/categorical_anomaly.py]: since `open` is a pretty generic function (also conincides with a Python built-in one), why not always use it as `rasterio.open` (as you do below) so that it doesn't accidentally get used in place of intended built-in `open`

File: recipes/categorical_anomaly.py [hunk 3]
```diff
@@ -5,18 +5,23 @@
 import geopandas as gpd
 import tqdm
 
+import rasterio
+import rasterio.mask
```
Comment (woozyking) [inline recipes/categorical_anomaly.py:9]: based on the usage pattern below, this import might be redundant

File: recipes/categorical_anomaly.py [hunk 4]
```diff
@@ -123,6 +139,26 @@ def tune_params(self,df, metrics_column):
         print("\nTuned LOF Parameters : {}".format(self.tuned_params))
         return(self.tuned_params)
 
+    def raster_to_vector(self,):
```
Comment (woozyking) [inline recipes/categorical_anomaly.py]: if there's no non-object arguments, remove the trailing `,` as `def raster_to_vector(self):`

on the other hand, this is a great place to apply method-level flexibility so that one can either initialize the object with a `filepath`, or apply one when running the method -- since conceptually there's no other coupling of the geotif file than this particular method 

```python
def raster_to_vector(self, geotif_path):
    fpath = geotif_path or self.geotif_path
```

also notice the suggestion to rename `self.filepath` to `self.geotif_path` (or something along that line) to better represent its intent than an overly generic term

File: recipes/categorical_anomaly.py [hunk 5]
```diff
@@ -123,6 +139,26 @@ def tune_params(self,df, metrics_column):
...
+        """
+        Given a filepath to a geotif file, will match the first layer of the raster tif to the
+        vector geometries located in the geodataframe given by the user.
+        Ex:
+        https://ghsl.jrc.ec.europa.eu/download.php?ds=pop
+        Using this dataset we can match the population geotif data with the FSA geometries in a geocohort,
+        and generate the population for each FSA
+        """
+        src = rasterio.open(self.filepath)
+        region = src.read(1, window=src.window(*self.df.geometry.total_bounds))
```
Comment (woozyking) [inline recipes/categorical_anomaly.py]: prefer the use of context manager whenever available, so in case of exception the I/O connection is always handled for you:

```python
with rasterio.open(fpath) as src:
    ...
```

File: recipes/categorical_anomaly.py [hunk 6]
```diff
@@ -123,6 +139,26 @@ def tune_params(self,df, metrics_column):
...
+        region[region<0] = None
```
Comment (woozyking) [inline recipes/categorical_anomaly.py]: how is this `region` used?

File: recipes/categorical_anomaly.py [hunk 7]
```diff
@@ -123,6 +139,26 @@ def tune_params(self,df, metrics_column):
...
+
+        def get_population_count(vector_polygon,raster_layer):
+            gtraster, bound = rasterio.mask.mask(raster_layer, [vector_polygon], crop=True)
+            pop_estimate = gtraster[0][gtraster[0]>0].sum()
+            return (pop_estimate.round(2))
```
Comment (woozyking) [inline recipes/categorical_anomaly.py]: redundant outer parentheses

File: recipes/categorical_anomaly.py [hunk 8]
```diff
@@ -123,6 +139,26 @@ def tune_params(self,df, metrics_column):
...
+        Using this dataset we can match the population geotif data with the FSA geometries in a geocohort,
+        and generate the population for each FSA
+        """
+        src = rasterio.open(self.filepath)
+        region = src.read(1, window=src.window(*self.df.geometry.total_bounds))
+        region[region<0] = None
+
+        def get_population_count(vector_polygon,raster_layer):
+            gtraster, bound = rasterio.mask.mask(raster_layer, [vector_polygon], crop=True)
+            pop_estimate = gtraster[0][gtraster[0]>0].sum()
```
Comment (woozyking) [inline recipes/categorical_anomaly.py]: leave space around operators like `>` so:

```python
pop_estimate = gtraster[0][gtraster[0] > 0].sum()
```

you can use [flake8 ](https://flake8.pycqa.org/en/latest/) or [black](https://github.com/psf/black) for guide/auto-format options

File: recipes/categorical_anomaly.py [hunk 9]
```diff
@@ -141,15 +177,24 @@ def create_model(self,auto_tune=None, **model_args):
...
-        score_df = df.copy()
-        if self.model_type == 'LODA':
+        try:
+            score_df = self.scaled_df.copy()
+        except:
+            score_df = self.df.copy()
+        if self.model_type == 'LODA' and (type(self.metrics_column) == list):
+            score_df['score'] = self.model.fit_predict(score_df[self.metrics_column])
+            score_df['anomaly'] = score_df['score'] <= -1
+            self.score_df = score_df
```
Comment (woozyking) [inline recipes/categorical_anomaly.py:225]: this assignment should happen after both conditionals (or the below `elif` becomes void)

File: recipes/categorical_anomaly.py (hunk 2 above)
Comment (KrisCowie) [reply recipes/categorical_anomaly.py]: Ah ya good point, I think I'd maybe gotten a little frustrated when some rasterio function wasn't working with just a import rasterio so I might've gone a little over board on the importing, and didn't clean them up perfectly. Will change!

File: recipes/categorical_anomaly.py (hunk 4 above)
Comment (KrisCowie) [reply recipes/categorical_anomaly.py]: I was wondering about that, I initially had it as just a variable within the function you called, then I thought maybe the self variable would be better? 
But without it being used elsewhere, and as something else the user would have to remember upon calling the class, you're probably right, just do it within (and more specific for the name! :) )

File: recipes/categorical_anomaly.py (hunk 1 above)
Comment (KrisCowie) [reply recipes/categorical_anomaly.py]: Ah, yup, sorry, this bug took me a while to figure out, I noticed sometimes metric column would have 'id' and 'anomaly' in it, figured out it came after calling graph_geo because I was appending to a variable I was pointing at self.metrics_column, figured I'd drop .copy() on both to be safe. I'll remove the top one.

File: recipes/categorical_anomaly.py (hunk 6 above)
Comment (KrisCowie) [reply recipes/categorical_anomaly.py]: Ah that's from when I was doing some mapping of the populations, shouldn't still be in there (reading turns it into a numpy array, and the original tif had -200's instead of 0s where there was no pop, hence the conversion). 
I could choropleth mapping as an option for graph_geo?

File: recipes/categorical_anomaly.py (hunk 9 above)
Comment (KrisCowie) [reply recipes/categorical_anomaly.py:225]: I couldn't get this to work? Tried moving the assignment but then it would fail when you were using say Loda and a list of metrics, or it wouldn't work with a single metric in a list. Just changed the logic in the if statements, it's working now? I tried every combination - LODA, one metric (str, and list), LODA multiple metrics, others one metric (string and list), and others multiple metrics, and it seems to be working

File: recipes/categorical_anomaly.py (hunk 2 above)
Comment (woozyking) [reply recipes/categorical_anomaly.py]: if you configure editor with linting (e.g. vs code + python plugin) it should point out unused imports for you

File: recipes/categorical_anomaly.py (hunk 9 above)
Comment (woozyking) [reply recipes/categorical_anomaly.py:225]: weird, given all conditions conclude with the same assignment back to the object property

```python
self.score_df = score_df
```

File: recipes/categorical_anomaly.py [hunk 10]
```diff
@@ -124,6 +138,25 @@ def tune_params(self,df, metrics_column):
...
+        Given a filepath to a geotif file, will match the first layer of the raster tif to the
+        vector geometries located in the geodataframe given by the user.
+        Ex:
+        https://ghsl.jrc.ec.europa.eu/download.php?ds=pop
+        Using this dataset we can match the population geotif data with the FSA geometries in a geocohort,
+        and generate the population for each FSA
+        """
+        fpath = geotif_path or self.geotif_path
+        with rasterio.open(fpath) as src:
+            def get_population_count(vector_polygon,raster_layer):
```
Comment (woozyking) [inline recipes/categorical_anomaly.py]: prefer space after `,`. consider `flake8` or `black` to catch/auto-format these for you

File: recipes/categorical_anomaly.py [hunk 11]
```diff
@@ -124,6 +138,25 @@ def tune_params(self,df, metrics_column):
...
+                gtraster = rasterio.mask.mask(raster_layer, [vector_polygon], crop=True)
+                pop_estimate = gtraster[0][gtraster[0] > 0].sum()
+                return pop_estimate.round(2)
```
Comment (woozyking) [inline recipes/categorical_anomaly.py]: the closure doesn't have to live within the context `with` right? or even better, can it be a pure function declared outside of the class altogether?

File: recipes/categorical_anomaly.py [hunk 12]
```diff
@@ -124,6 +138,25 @@ def tune_params(self,df, metrics_column):
...
+            self.df['pop_count'] = self.df['geometry'].apply(get_population_count,raster_layer=src)
```
Comment (woozyking) [inline recipes/categorical_anomaly.py]: prefer space after `,` based on https://pep8.org/

File: recipes/categorical_anomaly.py [hunk 13]
```diff
@@ -171,40 +213,60 @@ def get_anomalies(self):
...
+        Allows for both 2d and 3d PCA, and is typically helpful once the metrics columns are > 3
+
+        The graph's hover data will show the non-scaled metrics_column, as scaled metrics / PCA numbers aren't generally helpful to users.
+        The size of each point is reflective of how anomalous that point is - LOF uses a score called negative_outlier_factor to score how anomalous a point is,
+        while LODA uses score_samples, which need to be rescaled to only positive integers.
+        """
         try:
             if three_d:
-                if self.model_type == 'LODA':
+                if self.model_type == 'LODA' and (type(self.metrics_column) != list):
```
Comment (woozyking) [inline recipes/categorical_anomaly.py]: use `type(...) is list` (or in this case `type(...) is not list`) instead

**Review by woozyking — APPROVED** [review]
looks good

tho the graphs don't seem to show up even using nbviewer: https://nbviewer.org/github/EQWorks/locussdk-collab/blob/raster-vector/pop_and_scaling_ex.ipynb

**Comment by woozyking** [comment]
also lift the PR up from draft mode when you think it's ready for review

**Comment by KrisCowie** [comment]
Alright, just did one final commit with a normalization function, so users can select a field to first normalize their data around, and then scale it appropriately - so in our example, we can enter population count, it will divide each of the numeric features by that pop_count, and produce per capita values. Then it continues as normal

File: recipes/categorical_anomaly.py [hunk 14]
```diff
@@ -163,6 +164,21 @@ def raster_to_vector(self, geotif_path=None):
             return self.df
 
 
+    def normalize(self, field_1 = None,):
```
Comment (woozyking) [inline recipes/categorical_anomaly.py]: why is it named as `field_1` instead of just `field` (which correlates better with the docstring "Given a field..."?

redundant trailing `,` for a one-line definition

typically tho, it's a good thing to keep only when you utilize multiple lines for easier diffs:

```python
def normalize(
    self,
    field_1,
)
```

this way future changes would be isolated to single-line changes

Comment (KrisCowie) [reply recipes/categorical_anomaly.py]: Ah because initially I had the option to normalize by two fields - ex. population, then area to give a clicks per person per area sort of thing, but then Steven kind of mentioned that might not be a good idea, so I removed the option for the second level of normalization, but kept the variable name. Haha I can change it back

Comment (woozyking) [reply recipes/categorical_anomaly.py]: i see. reasonable, as long as it's documented so (especially for ppl who tend to always provide both `key=value` in function/method calls)

**Review by MadisonKonig — APPROVED** [review]
Looks good to me!

File: recipes/categorical_anomaly.py [hunk 15]
```diff
@@ -72,7 +96,7 @@ def tune_params(self,df, metrics_column):
                 else:
                     clf.fit_predict(df[[metrics_column]])
                 X_scores = np.log(- clf.negative_outlier_factor_)
-                t0 = X_scores.argsort()#[::-1]
+                t0 = X_scores.argsort()  # [::-1]
                 top_k = t0[-samps:]
                 min_k = t0[:samps]
```
Comment (QingleiZhang) [inline recipes/categorical_anomaly.py]: from the paper, the `M_k,in`, `V_k,in` are calculated by the normal points ((inliers) which have the highest local outlier factor scores, but looks like `t0[:samps]` defined those normal points with the lowers local outlier factor scores?

File: recipes/categorical_anomaly.py [hunk 16]
```diff
@@ -124,45 +147,94 @@ def tune_params(self,df, metrics_column):
         return(self.tuned_params)
```
Comment (QingleiZhang) [inline recipes/categorical_anomaly.py:148]: why don't put the value assignments of lines `144~145` under the condition `140` after line `141` directly?  The comparison of line `143` seems redundant here.

File: recipes/categorical_anomaly.py [hunk 17]
```diff
@@ -124,45 +147,94 @@ def tune_params(self,df, metrics_column):
...
-        if self.model_type == 'LODA':
+        try:
+            score_df = self.scaled_df.copy()
+        except:
+            score_df = self.df.copy()
+        if self.model_type == 'LODA' and ((type(self.metrics_column) is list) and len(self.metrics_column) > 1):
+            score_df['score'] = self.model.fit_predict(score_df[self.metrics_column])
+            score_df['anomaly'] = score_df['score'] <= -1
+            self.score_df = score_df
+        elif self.model_type == 'LODA' and type(self.metrics_column) is not list or (type(self.metrics_column) is list and len(self.metrics_column) < 2):
```
Comment (QingleiZhang) [inline recipes/categorical_anomaly.py]: do you miss parentheses for the rest after `self.model_type == 'LOAD'` ?   should be `elif self.model_type == 'LODA' and (type(self.metrics_column) is not list or (type(self.metrics_column) is list and len(self.metrics_column) < 2))`

File: recipes/categorical_anomaly.py [hunk 18]
```diff
@@ -228,11 +333,16 @@ def graph_highd(self, three_d: bool = None, **kwargs):
...
                     components = pca.fit_transform(feat_vals)
                     loadings = pca.components_.T * np.sqrt(pca.explained_variance_)
-                    score_df = pd.concat([self.score_df, pd.DataFrame(components)], axis=1)
+                    score_df = pd.concat([self.score_df.reset_index(), pd.DataFrame(components)], axis=1).reset_index()
                     score_df = score_df.rename(columns={0:'PC1', 1:'PC2'})
 
-                    fig = px.scatter(score_df, x='PC1', y='PC2', title='PCA Plot - Anomalies',
+                    # Append values to non-scaled df so users can see
+                    # the normal values for their features graphed
+                    scores = self.df.reset_index().join(score_df[['score', 'anomaly', 'PC1', 'PC2',]])
```
Comment (QingleiZhang) [inline recipes/categorical_anomaly.py]: Looks like line `341~362` are  same as line `309~329` (except line `315` and `348`), could we move part of the else condition clause (line 332-336) above the line 309 to make the code more concise

File: recipes/categorical_anomaly.py (hunk 15 above)
Comment (KrisCowie) [reply recipes/categorical_anomaly.py]: Oh you're right! It should've been:
min_k = t0[(2*-samps):-samps]
Basically taking the top [cn] predicted normal points. 

And should rename those to something like:
cn_outliers
cn_top_inliers

Really good catch! Thanks!

File: recipes/categorical_anomaly.py (hunk 17 above)
Comment (KrisCowie) [reply recipes/categorical_anomaly.py]: ~~I think it's right? It's supposed to cover two options:
model_type LODA and metrics column is not list
OR
model_type LODA and metrics_column is list, and len(metrics column <2)
LODA can't be used  with only one metric, so it's supposed to cover if a user is either a) using it with only one metric, and they've typed that one metric as a string, or b) using it with only one metric, but they've put it into a list for some reason~~

Oh I see what you mean, if it's model_type != LODA, and metrics_column = list, len <2 , it will identify as True, when it should only be on model_type == LODA. 
Haha you're right again!

File: recipes/categorical_anomaly.py (hunk 18 above)
Comment (KrisCowie) [reply recipes/categorical_anomaly.py]: Ya I think I was having trouble getting a simple way around that numeric_functions to metrics_column thing, it's such an edge case, but once I started with the logic to include it, I needed it in every function.
Just changed it to make it more concise by doing an if statement, then changing self.metrics_column to the numeric features if that logic is triggered.

File: recipes/categorical_anomaly.py (hunk 16 above)
Comment (KrisCowie) [reply recipes/categorical_anomaly.py:148]: Good point, changed

**Review by QingleiZhang — APPROVED** [review]
